src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+  bool caughtException = false;
+  if (rpcResult.hasException()) {
+    caughtException = true;
+    serialized = rpcResult.getException().getV8Serialized().asBytes();
+  } else {
+    serialized = rpcResult.getValue().getV8Serialized().asBytes();
+  }
+  auto deserialized = deserializeV8(js, kj::heapArray(KJ_REQUIRE_NONNULL(serialized).asBytes()));
+  if (caughtException) {
+    // The remote Worker threw an exception, so we should throw on the client side too.
+    js.throwException(deserialized);
+  }
+
+  KJ_REQUIRE(customEvent.outcome == EventOutcome::OK, "got unexpected event outcome for WorkerRpc");
+  return jsg::Value(js.v8Isolate, kj::mv(deserialized));
+}
+
 kj::Maybe<jsg::JsValue> WorkerRpc::getNamed(jsg::Lock& js, kj::StringPtr name) {
-    // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
-    //             this is just a temporary placeholder for testing.
-    return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
-        [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
-          return js.error("WorkerRpc is unimplemented");
+  // Named intercept is enabled, this means we won't default to legacy behavior.
+  // The return value of the function is a promise that resolves once the remote returns the result
+  // of the RPC call.
+  return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+      (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [this] (jsg::Lock& js, WorkerRpc::ResponseAndResult result) -> jsg::Value {
+          return handleRpcResponse(js, kj::mv(result));
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<void>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    // Try to execute the requested method.
+    try {
+      co_await ctx.run(kj::coCapture(
+          [this, methodName=kj::mv(methodName), serializedArgs=kj::mv(serializedArgs),
+           entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+              kj::Promise<void> {
+
+        jsg::Lock& js = lock;
```

Comment:
We are running in a coroutine, but you are storing a reference to `jsg::Lock` in a local variable. That reference becomes invalid after the first `co_await` -- and indeed, any code executing after the first `co_await` no longer runs with the isolate lock.

Looking at code below, it seems you are also storing `v8::Local` and similar objects on in local variables in coroutines. It's not safe to store these types past the end of the `v8::HandleScope`, which similarly goes away when we return to the event loop.

In general, mixing coroutines and calls to JS APIs just isn't safe -- or put another way, any use of a `jsg::Lock` cannot be in a coroutine. I think you will need to convert some of this code away from coroutines, unfortunately.

Reply:
https://github.com/cloudflare/workerd/pull/1028#discussion_r1412626362